nano/secure/rep_weights.cpp

```diff
@@ -0,0 +1,139 @@
+#include <nano/secure/rep_weights.hpp>
+#include <nano/store/component.hpp>
+#include <nano/store/rep_weight.hpp>
+
+nano::rep_weights::rep_weights (nano::store::rep_weight & rep_weight_store_a) :
+	rep_weight_store{ rep_weight_store_a }
+{
+}
+
+void nano::rep_weights::representation_add (store::write_transaction const & txn_a, nano::account const & rep_a, nano::uint128_t const & amount_a)
+{
+	auto previous_weight{ rep_weight_store.get (txn_a, rep_a) };
+	auto new_weight = previous_weight + amount_a;
+	put_store (txn_a, rep_a, previous_weight, new_weight);
+	nano::lock_guard<nano::mutex> guard (mutex);
+	put_cache (rep_a, new_weight);
+}
+
+void nano::rep_weights::representation_add_dual (store::write_transaction const & txn_a, nano::account const & rep_1, nano::uint128_t const & amount_1, nano::account const & rep_2, nano::uint128_t const & amount_2)
+{
+	if (rep_1 != rep_2)
+	{
+		auto previous_weight_1{ rep_weight_store.get (txn_a, rep_1) };
```

Comment:
Can this just be written as two calls to representation_add?

Reply:
The issue here is with the lock. Two calls to representation add would cause two separate locks. We could introduce a helper function `representation_add_locked`, but then the LMDB table write would be inside the lock, which I think is not optimal, because the lock would be held longer.